cpp/unreal_plugins/SpServices/Source/SpServices/EnhancedInputService.h

```diff
@@ -192,14 +194,41 @@ class EnhancedInputService : public Service {
                 }
             });
 
-        unreal_entry_point_binder->bindFuncToExecuteOnGameThread("enhanced_input_service", "inject_debug_key_for_actor",
-            [this](
-                uint64_t& actor,
-                std::string& chord_string,
-                std::string& key_event_string,
-                std::string& input_action_value_string,
-                bool verbose) -> void {
+        unreal_entry_point_binder->bindFuncToExecuteOnGameThread(
```

Comment:
I don't think I'm sold on the approach in this `inject_input_for_blueprint_actor` function yet. I like that it can work on actors that aren't currently possessed, but I don't like that it only works for BP types.

In contrast, I think it would be better for a user that wants to trigger actions on a specific un-possessed actor to simply possess the actor via the following approach:
- Get the currently possessed pawn using our `world_service.get_first_player_controller()` function and the `AController::K2_GetPawn()` function, which is a `UFUNCTION`.
- Unpossess the currently possessed pawn using `AController::UnPossess()`, which is a `UFUNCTION`.
- Possess their desired new pawn by `AController::Possess(APawn* InPawn)`, which is also a `UFUNCTION`.
- Trigger the desired actions using our `enhanced_input_service.inject_input_for_actor(...)` function.
- Unpossess the new pawn and re-possess the old pawn to restore everything to its original state.

I prefer this approach because it keeps our C++ smaller and simpler, and will work for both C++ and BP types.

An alternative approach that would work for C++ types would be to create our own custom `UInputComponent` type that we could pass into a pawn's SetupPlayerInputComponent(...) to intercept all the pawn's `Bind(...)` calls. But this is basically what `AController::Possess(...)` does, so I think it's cleaner to just embrace the fact that we can possess and un-possess using our existing API, and to require the user to possess their desired pawn.

More generally, if there is some chunk of application-specific functionality that requires chaining several Unreal function calls together, and that functionality can be expressed using existing primitives in our API, then we need an especially good reason to implement that functionality in our C++ code, as opposed to requiring the user to implement it themselves using our existing primitives.

The goal of SPEAR not to maximize the Unreal Engine's ease-of-use by providing high-level functions. The goal of SPEAR is to maximize the amount of Unreal functionality that can be expressed in Python by providing low-level primitives that compose together effectively.

Reply:
- I agree that the user couldn't do all of the steps in a single `with begin_frame()` block. Instead, the user would have to do some of the steps in `with begin_frame()` and some of the steps in the matching `with end_frame()`, but I think that is ok. It seems like a sensible use of the `begin_frame()` / `end_frame()` programming model.
- I think you're right that one controller can only possess one pawn. I had previously assumed this wouldn't be an issue, but admittedly I didn't look into it. I think you're right though.
- It would be nice to be able to control multiple unpossessed pawns, and it would be annoying if the user needed to create a new controller for each pawn they wanted to control. Maybe implementing a custom `UInputComponent` and passing it to `SetupPlayerInputComponent(...)` every time we want to trigger something isn't such a bad idea. If we did that, our approach would work for C++ and Blueprint types, and we could control multiple unpossessed pawns, and we wouldn't burden the user with implementing all of the possess/un-possess/re-possess logic.

Reply:
As we discussed earlier, SetupPlayerInputComponent is a protected function which cannot be called directly. Based on my test, static_cast or reinterpret_cast to ASpPawn would not call the function we want. Another work around would be let PlayerController possess the pawn, and then store or return the function pointers.

Reply:
This code works fine for me on macOS. Does this not work for you?

Note that I just hacked this test code into `SpGameMode.cpp` because it was the most convenient place to experiment with a valid APawn pointer. I'm not suggesting that we actually check in code like this.

```
//
// SpGameMode.cpp
//

// We don't use the usual Unreal naming convention here because PawnWrapper is not a UCLASS.
class PawnWrapper : public APawn
{
public:
    void callSetupPlayerInputComponent(UInputComponent* input_component)
    {
        SP_LOG_CURRENT_FUNCTION();
        SetupPlayerInputComponent(input_component);
    }
};

void ASpGameMode::PostLogin(APlayerController* new_player)
{
    AGameModeBase::PostLogin(new_player);

    // Set the stable name for the DefaultPawnClass instance so we can find it later. We only do this if the
    // pawn is non-null (it is possible for pawn to be null if we press "Simulate" in the editor), and if the
    // pawn has a USpStableNameComponent.
    APawn* pawn = new_player->GetPawn();
    if (pawn && Unreal::hasStableName(pawn)) {
        Unreal::setStableName(pawn, "__SP_DEFAULT_PAWN__");

        SP_LOG_CURRENT_FUNCTION();
        PawnWrapper* pawn_wrapper = static_cast<PawnWrapper*>(pawn); // this cast is unsafe because the runtime type of pawn is not PawnWrapper
        pawn_wrapper->callSetupPlayerInputComponent(nullptr);
    }
}

//
// SpSpectatorPawn.h
//

UCLASS()
class ASpSpectatorPawn : public ASpectatorPawn
{
    GENERATED_BODY()
public:

    // ...
    void SetupPlayerInputComponent(UInputComponent* input_component) override;
    // ...
};

//
// SpSpectatorPawn.cpp
//

void ASpSpectatorPawn::SetupPlayerInputComponent(UInputComponent* input_component)
{
    SP_LOG_CURRENT_FUNCTION();

    if (input_component) {
        ASpectatorPawn::SetupPlayerInputComponent(input_component);
    }
}
```

Reply:
It could be the case that I did not implemented it correctly. I though we do not know the exact pawn class at runtime?

Reply:
Tested your code and it works. 

Based on this, I I implemented `setup_enhanced_input_component` that call `CreateInputComponent` and `SetupPlayerInputComponent`. This allows us to control multi-agents very cleanly for both bp type and cpp type. 

I tried the pawn from `TP_ThirdPerson` example. Although `TP_ThirdPersonCharacter::Move` is called, it checks if its controller pointer is valid, which may require some extra setup for multi-agent control. Nevertheless, I think our code is general enough and `EnhancedInputService` should not take care of this speical case.

Reply:
Maybe we could have a multi-agent example, especially when it requires very little extra work.